### ISSUE TITLE
Data compress/decompress NOT match

### ISSUE THREAD

**Issue by invxp** [zhaihj/lz77-rust#1]
fn test()
{
    //sampling data
    let mut data = [0u8; 27];
    data[0] = 185;
    data[1] = 254;
    data[2] = 185;
    data[3] = 254;
    data[4] = 185;
    data[5] = 254;
    data[6] = 185;
    data[7] = 254;
    data[8] = 33;
    data[9] = 33;
    data[10] = 33;
    data[11] = 43;
    data[12] = 42;
    data[13] = 35;
    data[14] = 38;
    data[15] = 42;
    data[16] = 40;
    data[17] = 35;
    data[18] = 38;
    data[19] = 40;
    data[20] = 42;
    data[21] = 35;
    data[22] = 38;
    data[23] = 40;
    data[24] = 42;
    data[25] = 35;
    data[26] = 64;

    let mut comp = Vec::new();
    lz77_compress(&data, &mut comp);
    let mut dec = Vec::new();
    lz77_decompress(&comp, &mut dec);

    //assertion raised!!!!!!!
    assert!(comp.len() < data.len());
    assert_eq!(data.to_vec(), dec);
}

**Comment by zhaihj** [comment]
TL;DR

check `experiment` branch.

---


In your example, compress&decompress matches the raw. 
The assertion fails because compressed data is bigger than raw.
You can just comment that line out.

You may wonder why the compressed data becomes bigger, here is a manual analysis:

`[185 ,254 ,185 ,254 ,185 ,254 ,185 ,254 ,33
,33 ,33 ,43 ,42 ,35 ,38 ,42 ,40 ,35 ,38 ,40
,42 ,35 ,38 ,40 ,42 ,35 ,64]`

THRESHOLD is 3, so the first compression happens at `data[4]`.
Manually compression gives:

`[flag, 185 ,254 ,185 ,254 ,pos, len ,33
,33 ,33 , flag, 43 ,42 ,35 ,38 ,42 ,40 ,35 ,38 flag,
,40 ,pos, len ,pos, len ,64, 0, 0, 0, 0]`

Last 4 zero-valued bytes. In current algoritm, to make sure the stream is correctly ended,
we add another extra 3 bytes to the end, so finally it becomes:

`[flag, 185 ,254 ,185 ,254 ,pos, len ,33
,33 ,33 , flag, 43 ,42 ,35 ,38 ,42 ,40 ,35 ,38 flag,
,40 ,pos, len ,pos, len ,64, 0, 0, 0, 0, 0, 0, 0]`

If you really care about the size of compressed stream, 
there is a un-fully-tested version in `experiment`.


Here is the compilabse source code:
```rust
extern crate lz77;
use lz77::lz77::*;

fn main() {
let data = [185 ,254 ,185 ,254 ,185 ,254 ,185 ,254 ,33
            ,33 ,33 ,43 ,42 ,35 ,38 ,42 ,40 ,35 ,38 ,40
            ,42 ,35 ,38 ,40 ,42 ,35 ,64];
let mut comp = Vec::new();
lz77_compress(&data, &mut comp);
let mut dec = Vec::new();
lz77_decompress(&comp, &mut dec);

// assertion raised because compressed data is bigger than raw.
// you can just comment out it.
//assert!(comp.len() < data.len());

// this assert works well.
assert_eq!(data.to_vec(), dec);
}
```rust

**Comment by zhaihj** [comment]
IMO, this issue has been solved. 
If you have other problems, another issue or pull request is welcomed.
